File: java/src/main/java/sk/epostak/sdk/resources/InboxResource.java
package sk.epostak.sdk.resources;

import sk.epostak.sdk.HttpClient;
import sk.epostak.sdk.models.*;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Access received (inbound) documents.
 * <p>
 * Provides listing, detail retrieval, and acknowledgement of documents
 * received via Peppol. Integrator keys can also list documents across all firms.
 * <p>
 * Access via {@code client.documents().inbox()}.
 *
 * <pre>{@code
 * // List new documents, then acknowledge each one
 * InboxListResponse inbox = client.documents().inbox().list(null, 50, "RECEIVED", null);
 * for (Document doc : inbox.documents()) {
 *     process(doc);
 *     client.documents().inbox().acknowledge(doc.getId());
 * }
 * }</pre>
 */
public final class InboxResource {

    private final HttpClient http;

    /**
     * Creates a new inbox resource.
     *
     * @param http the HTTP client used for API communication
     */
    InboxResource(HttpClient http) {
        this.http = http;
    }

    /**
     * List received documents with filtering and pagination.
     *
     * @param offset pagination offset (0-based), or {@code null} for default
     * @param limit  max results per page (1-100, default 20), or {@code null} for default
     * @param status filter by status: {@code "RECEIVED"} or {@code "ACKNOWLEDGED"}, or {@code null} for all
     * @param since  only return documents received after this ISO 8601 timestamp, or {@code null}
     * @return paginated list of inbox documents
     * @throws sk.epostak.sdk.EPostakException if the request fails
     */
    public InboxListResponse list(Integer offset, Integer limit, String status, String since) {
        Map<String, Object> params = new LinkedHashMap<>();
        params.put("offset", offset);
        params.put("limit", limit);
        params.put("status", status);
        params.put("since", since);
        return http.get("/documents/inbox" + HttpClient.buildQuery(params), InboxListResponse.class);
    }

    /**
     * List received documents with default parameters.
     *
     * @return paginated list of inbox documents (offset 0, limit 20, all statuses)
     * @throws sk.epostak.sdk.EPostakException if the request fails
     */
    public InboxListResponse list() {
        return list(null, null, null, null);
    }

    /**
     * Get full detail of an inbox document including the UBL XML payload.
     *
     * @param id the inbox document UUID
     * @return the document detail with UBL XML content
     * @throws sk.epostak.sdk.EPostakException if the document is not found or the request fails
     */
    public InboxDocumentDetailResponse get(String id) {
        return http.get("/documents/inbox/" + HttpClient.encode(id), InboxDocumentDetailResponse.class);
    }

    /**
     * Mark an inbox document as processed (acknowledged). This is an idempotent
     * operation -- acknowledging an already-acknowledged document is a no-op.
     *
     * @param id the inbox document UUID
     * @return the acknowledgement confirmation with timestamp
     * @throws sk.epostak.sdk.EPostakException if the document is not found or the request fails
     */
    public AcknowledgeResponse acknowledge(String id) {
        return http.post("/documents/inbox/" + HttpClient.encode(id) + "/acknowledge", null, AcknowledgeResponse.class);
    }

    /**
     * List received documents across all firms. Requires an integrator key ({@code sk_int_*}).
     *
     * @param offset pagination offset (0-based), or {@code null} for default
     * @param limit  max results per page (1-200, default 50), or {@code null} for default
     * @param status filter by status: {@code "RECEIVED"} or {@code "ACKNOWLEDGED"}, or {@code null} for all
     * @param since  only return documents received after this ISO 8601 timestamp, or {@code null}
     * @param firmId filter to a specific firm UUID, or {@code null} for all firms
     * @return paginated list of documents across all firms
     * @throws sk.epostak.sdk.EPostakException if the request fails or the key is not an integrator key
     */
    public InboxAllResponse listAll(Integer offset, Integer limit, String status, String since, String firmId) {
        Map<String, Object> params = new LinkedHashMap<>();
        params.put("offset", offset);
        params.put("limit", limit);
        params.put("status", status);
        params.put("since", since);
        params.put("firm_id", firmId);
        return http.get("/documents/inbox/all" + HttpClient.buildQuery(params), InboxAllResponse.class);
    }

    /**
     * List received documents across all firms with default parameters.
     *
     * @return paginated list of documents across all firms (offset 0, limit 50, all statuses)
     * @throws sk.epostak.sdk.EPostakException if the request fails or the key is not an integrator key
     */
    public InboxAllResponse listAll() {
        return listAll(null, null, null, null, null);
    }
}

File: dotnet/src/EPostak/Resources/InboxResource.cs
using EPostak.Models;

namespace EPostak.Resources;

/// <summary>
/// Access received (inbound) documents from the Peppol network.
/// Provides listing, retrieval, and acknowledgement of incoming invoices
/// and other business documents.
/// </summary>
public sealed class InboxResource
{
    private readonly HttpRequestor _http;

    internal InboxResource(HttpRequestor http) => _http = http;

    /// <summary>
    /// List inbox documents with optional filtering by status or date.
    /// Results are paginated -- use <c>Offset</c> and <c>Limit</c> to page through results.
    /// </summary>
    /// <param name="params">Optional filters: pagination, status (RECEIVED/ACKNOWLEDGED), and date cutoff.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Paginated list of inbox documents with total count.</returns>
    /// <example>
    /// <code>
    /// var inbox = await client.Documents.Inbox.ListAsync(new InboxListParams
    /// {
    ///     Status = InboxStatus.RECEIVED,
    ///     Limit = 10
    /// });
    /// foreach (var doc in inbox.Documents)
    ///     Console.WriteLine($"{doc.Number} from {doc.Supplier.Name}: {doc.Totals.WithVat} {doc.Currency}");
    /// </code>
    /// </example>
    public Task<InboxListResponse> ListAsync(InboxListParams? @params = null, CancellationToken ct = default)
    {
        var qs = HttpRequestor.BuildQuery(
            ("offset", @params?.Offset?.ToString()),
            ("limit", @params?.Limit?.ToString()),
            ("status", @params?.Status?.ToString()),
            ("since", @params?.Since));
        return _http.RequestAsync<InboxListResponse>(HttpMethod.Get, $"/documents/inbox{qs}", ct);
    }

    /// <summary>
    /// Get a single inbox document including the full UBL XML payload.
    /// Use this to retrieve the raw Peppol BIS 3.0 XML for processing in your system.
    /// </summary>
    /// <param name="id">The inbox document UUID.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>The document metadata and UBL XML payload (null if not yet available).</returns>
    /// <example>
    /// <code>
    /// var detail = await client.Documents.Inbox.GetAsync("doc_abc123");
    /// if (detail.Payload is not null)
    ///     File.WriteAllText("received_invoice.xml", detail.Payload);
    /// </code>
    /// </example>
    public Task<InboxDocumentDetailResponse> GetAsync(string id, CancellationToken ct = default)
        => _http.RequestAsync<InboxDocumentDetailResponse>(HttpMethod.Get, $"/documents/inbox/{Uri.EscapeDataString(id)}", ct);

    /// <summary>
    /// Acknowledge receipt of an inbox document. This marks the document as processed
    /// in your system. Acknowledged documents won't appear in unprocessed filters
    /// and won't trigger repeated webhook/queue events.
    /// </summary>
    /// <param name="id">The inbox document UUID to acknowledge.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Confirmation with the document ID, new status, and acknowledgement timestamp.</returns>
    /// <example>
    /// <code>
    /// var ack = await client.Documents.Inbox.AcknowledgeAsync("doc_abc123");
    /// Console.WriteLine($"Acknowledged at: {ack.AcknowledgedAt}");
    /// </code>
    /// </example>
    public Task<AcknowledgeResponse> AcknowledgeAsync(string id, CancellationToken ct = default)
        => _http.RequestAsync<AcknowledgeResponse>(HttpMethod.Post, $"/documents/inbox/{Uri.EscapeDataString(id)}/acknowledge", ct);

    /// <summary>
    /// List inbox documents across all firms managed by an integrator key.
    /// Only available with integrator API keys (<c>sk_int_*</c>). Returns documents
    /// from all assigned firms with the firm ID and name on each entry.
    /// </summary>
    /// <param name="params">Optional filters: pagination, status, date cutoff, and specific firm UUID.</param>
    /// <param name="ct">Cancellation token.</param>
    /// <returns>Paginated list of inbox documents across all firms with total count.</returns>
    /// <example>
    /// <code>
    /// // Get all unacknowledged documents across firms
    /// var all = await client.Documents.Inbox.ListAllAsync(new InboxAllParams
    /// {
    ///     Status = InboxStatus.RECEIVED,
    ///     Limit = 50
    /// });
    /// foreach (var doc in all.Documents)
    ///     Console.WriteLine($"[{doc.FirmName}] {doc.Number}: {doc.Totals.WithVat} {doc.Currency}");
    /// </code>
    /// </example>
    public Task<InboxAllResponse> ListAllAsync(InboxAllParams? @params = null, CancellationToken ct = default)
    {
        var qs = HttpRequestor.BuildQuery(
            ("offset", @params?.Offset?.ToString()),
            ("limit", @params?.Limit?.ToString()),
            ("status", @params?.Status?.ToString()),
            ("since", @params?.Since),
            ("firm_id", @params?.FirmId));
        return _http.RequestAsync<InboxAllResponse>(HttpMethod.Get, $"/documents/inbox/all{qs}", ct);
    }
}
